contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -294,6 +298,28 @@ change_key(pam_handle_t *pamh, const char *ds_name,
 		zfs_close(ds);
 		return (-1);
 	}
+	keyformat = zfs_prop_get_int(ds, ZFS_PROP_KEYFORMAT);
+	if (keyformat != ZFS_KEYFORMAT_NONE) {
+		if (zfs_prop_get(ds, ZFS_PROP_KEYLOCATION, prop_keylocation,
+		    sizeof (prop_keylocation), NULL, NULL, 0, B_TRUE) != 0) {
+			pam_syslog(pamh, LOG_ERR,
+			    "Failed to get keylocation for '%s'", ds_name);
+			return (-1);
+		}
+		// derive keyfile from uri scheme
+		if (strncmp(prop_keylocation, "file://", 7) == 0) {
+			fp = prop_keylocation + 7;
+			if ((file = fopen(fp, "r")) == NULL) {
+				pam_syslog(pamh, LOG_ERR,
+				    "Failed to access keyfile '%s'", fp);
+			}
+			(void) fclose(file);
+			if (remove(fp) != 0) {
```

Comment:
Why is key file being removed?

Reply:
The keylocation given 'file:///keyfile' will soon be replaced by ['prompt'](https://github.com/openzfs/zfs/pull/11165/files#diff-987991ce1c5894484f401f92065c417acc226f5f79d8d86fad8b4eb0c1fd6aa8R323), after that the key file become unnecessary.

Reply:
I do not find this behavior good. The keyfile might be used by others (eg. other datasets) as well. I think removing it is counter-intuitive as this is also not the behavior of `zfs change-key`.

Reply:
The use case here is a new created dataset in encryption mapped to a new added user, the `file:///keyfile` for the corresponded dataset is roughly a temporary file because upcoming `change_key` sets `keylocation` to `prompt` hence deleting the file is painless. 

Can we make this deletion optional part of `argv`? If necessary I can cross check if the keyfile is being referenced elsewhere among datasets.

Reply:
I don't really understand why you would need a keyfile for the initial setup of the user. The method I currently use to setup new users is this:
```
zfs create -o encryption=aes-256-gcm -o keyformat=passphrase -o keylocation="file://${tmpfile}" rpool/home/"$1"
rm -- "$tmpfile"
zfs set keylocation=prompt rpool/home/"$1"
adduser "$1"
# ... setup other things in the user's home directory
zfs unmount "rpool/home/$1"
zfs unload-key "rpool/home/$1"
```
This sets the key for the home-dataset directly to the initial password of the user, so there is never a need to remove a keyfile for a user. And I find it reasonable that the script that creates the password file has to delete it. Maybe your use-case is a bit different so it would probably be good, it you could describe it in greater detail if that solution does not fit.

I find automatically deleting the keyfile dangerous in any case (probably even if it would have to be enabled manually), as you can never be sure that the keyfile wouldn't be used elsewhere (maybe even outside of zfs, so cross-checking other datasets does not really help in my opinion) and deleting it will destroy the secret permanently. For reference, `zfs change-key` also has no option to automatically delete a keyfile.

Reply:
Understood the risky deletion of keyfile, I will have to restructure the work of creating the [keyfile](https://github.com/ubuntu/zsys/pull/173/files#diff-01dcdf16dabef7115ecb9bc6c27b27cc250e33c878125090ca967da5d66b2d22R340) if customized deletion in `pam_zfs_key` isn't safe, and remove the 2nd commit from this PR.

The use case here is that dataset was created in encryption with a tmpfile similarly to your flow, just is tied to [adduser](https://git.launchpad.net/ubuntu/+source/adduser/tree/adduser?h=applied/ubuntu/focal#n715). The tmpfile was not deleted right after its creation because `keylocation`'s change happens until `pam_zfs_key`'s `change_key()`. Early deletion may cause the encrypted dataset no longer usable.